File: mobile/android/src/main/kotlin/com/otravelz/android/data/network/adapter/DomainAdapters.kt
package com.otravelz.android.data.network.adapter

import com.otravelz.android.data.network.dto.StopNearbyDto
import com.otravelz.android.data.network.dto.WeatherResponseDto

sealed interface WeatherState {
    data class Available(
        val locationName: String,
        val temperatureC: Double,
        val condition: String,
        val advice: String?
    ) : WeatherState

    data class Unavailable(val reason: String = "Sensor readings currently unavailable") : WeatherState
}

fun WeatherResponseDto.toDomain(): WeatherState {
    val temp = current?.temperatureC
    val cond = current?.condition
    return if (temp != null && cond != null) {
        WeatherState.Available(
            locationName = current.locationName ?: locationName ?: "Odisha",
            temperatureC = temp,
            condition = cond,
            advice = current.advice
        )
    } else {
        WeatherState.Unavailable()
    }
}

sealed interface TransitStopTruth {
    data class Exact(
        val stopId: String,
        val name: String,
        val latitude: Double,
        val longitude: Double,
        val distanceM: Double?
    ) : TransitStopTruth

    data class LocalityOnly(
        val stopId: String,
        val name: String,
        val locality: String,
        val city: String
    ) : TransitStopTruth
}

fun StopNearbyDto.toDomain(): TransitStopTruth {
    val lat = latitude
    val lon = longitude
    val isCandidate = coordinateStatus?.lowercase() == "candidate"

    return if (!isCandidate && lat != null && lon != null) {
        TransitStopTruth.Exact(
            stopId = stopId,
            name = name,
            latitude = lat,
            longitude = lon,
            distanceM = distanceM
        )
    } else {
        TransitStopTruth.LocalityOnly(
            stopId = stopId,
            name = name,
            locality = locality ?: "Unknown",
            city = city ?: "Odisha"
        )
    }
}

File: mobile/ios/OTravelz/Networking/Adapters/DomainAdapters.swift
import Foundation

/// Domain representation of live weather conditions.
public enum WeatherState: Equatable, Sendable {
    case available(locationName: String, temperatureC: Double, condition: String, advice: String?)
    case cached(locationName: String, temperatureC: Double, condition: String, advice: String?, relativeTimeAgo: String)
    case unavailable(reason: String)
}

public extension WeatherResponseDTO {
    func toDomain() -> WeatherState {
        guard let current = current,
              let temp = current.temperatureC,
              let cond = current.condition else {
            return .unavailable(reason: "Sensor readings currently unavailable")
        }
        return .available(
            locationName: current.locationName ?? locationName ?? "Odisha",
            temperatureC: temp,
            condition: cond,
            advice: current.advice
        )
    }
}

/// Truth boundary representation of a transit stop location.
public enum TransitStopTruth: Equatable, Sendable {
    case exact(stopId: String, name: String, latitude: Double, longitude: Double, distanceM: Double?)
    case localityOnly(stopId: String, name: String, locality: String, city: String)
}

public extension StopNearbyDTO {
    func toDomain() -> TransitStopTruth {
        let isCandidate = coordinateStatus?.lowercased() == "candidate"
        if !isCandidate, let lat = latitude, let lon = longitude {
            return .exact(
                stopId: stopId,
                name: name,
                latitude: lat,
                longitude: lon,
                distanceM: distanceM
            )
        } else {
            return .localityOnly(
                stopId: stopId,
                name: name,
                locality: locality ?? "Unknown",
                city: city ?? "Odisha"
            )
        }
    }
}
